### PR TITLE
Add media-adapter for vips

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | none
| License | MIT
| Documentation PR | https://github.com/sulu/sulu-docs/pull/463

#### What's in this PR?

This PR add adapter for [rokka-io/imagine-vips](https://github.com/rokka-io/imagine-vips/issues).

See Blog post about it at the [liip blog ](https://www.liip.ch/de/blog/libvips-adapter-for-php-imagine)

#### Why?

It is an alternative for imagick and gd.

#### Example Installation on mac

~~~bash
brew install pkg-config
brew install vips
pecl install vips
composer require rokka/imagine-vips
~~~

#### To Do

- [x] Create a documentation PR

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -158,6 +158,7 @@
         <service id="sulu_media.adapter.gd" class="Imagine\Gd\Imagine" public="false"/>
         <service id="sulu_media.adapter.imagick" class="Imagine\Imagick\Imagine" public="false"/>
         <service id="sulu_media.adapter.gmagick" class="Imagine\Gmagick\Imagine" public="false"/>
+        <service id="sulu_media.adapter.vips" class="Imagine\Vips\Imagine" public="false"/>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Resources/config/services.xml]: service should only be registered when class is available

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 2]
```diff
@@ -191,7 +191,11 @@ public function convert(FileVersion $fileVersion, $formatKey, $imageFormat)
 
         // Set Interlacing to plane for smaller image size.
         if (1 == count($image->layers())) {
-            $image->interlace(ImageInterface::INTERLACE_PLANE);
+            try {
+                $image->interlace(ImageInterface::INTERLACE_PLANE);
+            } catch (\Exception $exception) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Isn't there a better exception we can catch?

**Comment by wachterjohannes** [comment]
@danrot @alexander-schranz fixed and ready to review again

File: src/Sulu/Bundle/MediaBundle/Resources/config/services.xml (hunk 1 above)
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Resources/config/services.xml]: @wachterjohannes this should be removed from this file

Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Resources/config/services.xml]: done

**Comment by danrot** [comment]
I would say the `rokka/imagine-vips` package should at least be added as suggestion, shouldn't it?

File: src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php [hunk 3]
```diff
@@ -270,10 +273,15 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('command.xml');
 
         if ('auto' === $config['adapter']) {
-            $container->setAlias(
-                'sulu_media.adapter',
-                'sulu_media.adapter.' . (class_exists('Imagick') ? 'imagick' : 'gd')
-            );
+            $adapter = 'gd';
+            if (class_exists('Imagine\Vips\Imagine')) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php]: Any reason not to use the `::class` constant`

File: src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php [hunk 4]
```diff
@@ -270,10 +273,15 @@ public function load(array $configs, ContainerBuilder $container)
...
+                $loader->load('services_imagine_vips.xml');
+                $adapter = 'vips';
+            } elseif (class_exists('Imagick')) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php]: Same as above

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 5]
```diff
@@ -191,7 +191,13 @@ public function convert(FileVersion $fileVersion, $formatKey, $imageFormat)
 
         // Set Interlacing to plane for smaller image size.
         if (1 == count($image->layers())) {
-            $image->interlace(ImageInterface::INTERLACE_PLANE);
+            try {
+                $image->interlace(ImageInterface::INTERLACE_PLANE);
+            } catch (\InvalidArgumentException $exception) {
+                // ignore unsupported Interface type
+            } catch (\RuntimeException $exception) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Is there no better exception? Same question for the above `catch`

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: @wachterjohannes It should be a `Imagine\Exception\RuntimeException`

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 6]
```diff
@@ -191,7 +191,13 @@ public function convert(FileVersion $fileVersion, $formatKey, $imageFormat)
 
         // Set Interlacing to plane for smaller image size.
         if (1 == count($image->layers())) {
-            $image->interlace(ImageInterface::INTERLACE_PLANE);
+            try {
+                $image->interlace(ImageInterface::INTERLACE_PLANE);
+            } catch (\InvalidArgumentException $exception) {
+                // ignore unsupported Interface type
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: this seems have nothing todo with the [vips](https://github.com/rokka-io/imagine-vips/blob/3039f6a27d481c4a61363b68a910a8110973bcd1/lib/Imagine/Vips/Image.php#L505) adapter what case does this then happen should maybe be not part of this PR?

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 5 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: @alexander-schranz thats not correct - the Vips Image throws a `\RuntimeException` because the function it is not implemented there.

https://github.com/rokka-io/imagine-vips/blob/0.9.0/lib/Imagine/Vips/Image.php#L514

what you have changed yesterday there 🙈 i will also catch the new exception here.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 6 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: not directly but thats a possible exception which is thrown in other adapters of imagine

Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Which does not automatically mean that we should simply catch and ignore. This way we won't even realize that an exception happened. Or is that in any way required in order to work?

Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: yes - if you have installed the vips adapter (or any other adapter) which does not implement interlace sulu will crash :) we have to ignore it - as we also do it on other places in this class

Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: But as @alexander-schranz already described `interlace` in that case throws a `RuntimeException` (which I also don't like, but that seems to be the way it is...) not an `InvalidArgumentException`, which could also have loads of other causes.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 5 above)
Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Btw, I am even not sure if they did that on purpose, because sometime they throw the `Imagine\Exception\RuntimeException` as @alexander-schranz mentioned... I think that might be even a bug in their adapter.

File: src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php (hunk 3 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php]: yes - phpstan will throw an error  in the CI/CD (https://travis-ci.org/sulu/sulu/jobs/634123652#L1272) and we cannot install the lib - because then we would have to install the extension `vips`

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 5 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: For clearification in version < 1.0 imagine. Not Implemented or supported function did throw a `Imagine\Exception\RuntimeException`. In >=1.0 they did add a new exception the `Imagine\Exception\NotSupportedException` which extends from `Imagine\Exception\RuntimeException`. Sadly the vips adapter didn't implement that correctly so I created a PR not sure if they gonna release it soon.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 6 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: OK will remove the catch

File: src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php (hunk 3 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/DependencyInjection/SuluMediaExtension.php]: If its just phpstan I would add it to ignore errors.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
 namespace Sulu\Bundle\MediaBundle\Media\ImageConverter;
 
+use Imagine\Exception\NotSupportedException;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Is it safe to use this also with versions of the imagine library not having this exception? If not we would have to rise the minimum version of it in `composer.json`. On the other hand I think it is not necessary to catch this exception separately, if we already catch the `RuntimeException` (hopefully soon the one from Imagine if the imagine-vips adapter releases a new version), which the `NotSupportedException` is inheriting from.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 5 above)
Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Let's wait for their answer in rokka-io/imagine-vips#7

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 7 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: have removed the `NotSupportedException ` from the catch clause

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 8]
```diff
@@ -203,6 +203,8 @@ public function convert(FileVersion $fileVersion, $formatKey, $imageFormat)
             }
         } catch (RuntimeException $exception) {
             // ignore exceptions here (some imagine adapter does not implement this)
+        } catch (\RuntimeException $exception) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Wasn't the reason we were waiting for the other merge, that we don't need that `catch` anymore?

**Comment by danrot** [comment]
I can confirm that this does not break using the old imagine adapters. However, I have a hard time finding out how to install `vips` including the php extension on my system :see_no_evil: 

@alexander-schranz Would be great if you could test that, and merge if the tests succeed.

**Comment by alexander-schranz** [comment]
@wachterjohannes at current state to support animated gifs we need change the following in the `modifyAllLayers` function:

```diff
diff --git a/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php b/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php
index ba29d16187..6cd2557cf4 100644
--- a/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php
+++ b/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php
@@ -16,6 +16,7 @@ use Imagine\Filter\Basic\Autorotate;
 use Imagine\Image\ImageInterface;
 use Imagine\Image\ImagineInterface;
 use Imagine\Image\Palette\RGB;
+use Imagine\Vips\Image as VipsImage;
 use Sulu\Bundle\MediaBundle\Entity\FileVersion;
 use Sulu\Bundle\MediaBundle\Entity\FormatOptions;
 use Sulu\Bundle\MediaBundle\Media\Exception\ImageProxyInvalidFormatOptionsException;
@@ -403,7 +404,10 @@ class ImagineImageConverter implements ImageConverterInterface

         if (count($layers)) {
             $countLayer = 0;
-            $image->layers()->coalesce();
+
+            if (!$image instanceof VipsImage) {
+                $image->layers()->coalesce();
+            }

             /** @var ImageInterface $temporaryImage */
             $temporaryImage = null;
@@ -411,7 +415,7 @@ class ImagineImageConverter implements ImageConverterInterface
                 ++$countLayer;
                 $layer = call_user_func($modifier, $layer);
                 if (1 === $countLayer) {
-                    $temporaryImage = $layer; // use first layer as main image
+                    $temporaryImage = clone $layer; // use first layer as main image
                 } else {
                     $temporaryImage->layers()->add($layer);
                 }

```

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 9]
```diff
@@ -401,15 +402,20 @@ private function modifyAllLayers(ImageInterface $image, callable $modifier)
 
         if (count($layers)) {
             $countLayer = 0;
-            $image->layers()->coalesce();
+
+            // coalesce currently not supported by the VipsAdapter:
+            // see: https://github.com/rokka-io/imagine-vips/issues/
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Issue number would also be nice :sweat_smile:

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 10]
```diff
@@ -401,15 +402,20 @@ private function modifyAllLayers(ImageInterface $image, callable $modifier)
...
+            if (!$image instanceof VipsImage) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: @alexander-schranz Didn't we say that the `coalesce` function should be a noop in the `Vips` adapter, because it seems it is not necessary? If the maintainers agree and would make a release soonish, I would be more comfortable with waiting until this is fixed before merging this, instead of having this `if` with a reference to a concrete adapter...

**Comment by wachterjohannes** [comment]
@alexander-schranz your fix breaks the behavior for GD see travis error https://travis-ci.org/wachterjohannes/sulu/jobs/634609962#L1005

**Comment by alexander-schranz** [comment]
@wachterjohannes can nothing todo with that changes for the gif as that only applies to multi layer images and the test which is failing is a jpeg and that part is never reached by a jpg image.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 9 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: @wachterjohannes https://github.com/rokka-io/imagine-vips/issues/8

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php [hunk 11]
```diff
@@ -399,17 +400,22 @@ private function modifyAllLayers(ImageInterface $image, callable $modifier)
             $layers = [];
         }
 
-        if (count($layers)) {
+        if (count($layers) > 1) {
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: Did debug and this will save us a lot of memory 40mb -> 8mb tested on the imagick adapter.

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: @danrot backported this to 1.6: https://github.com/sulu/sulu/pull/4968

File: composer.json [hunk 12]
```diff
@@ -151,6 +152,7 @@
     },
     "suggest": {
         "contao/imagine-svg": "^0.2.2 - Rendering raw svg formats for svg uploads",
+        "rokka/imagine-vips": "^0.9.1 - Rendering formats by using the libvips",
```
Comment (danrot) [inline composer.json]: Replace that with `^0.9.2`

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php (hunk 11 above)
Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: I would prefer to not have this change in here - then it is less likely that there'll be a conflict when I merge back, and I have less to test in this PR :-)

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: @danrot tests will fail then as the clone doesn't work on `GD` driver.

Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: But doesn't that mean that it will fail if you upload an animated GIF if using GD in general? :thinking:

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php]: GD doesn't support animated gifs or other multi layer image formats [see](https://imagine.readthedocs.io/en/stable/usage/layers.html#disclaimer).
